Pull request for creating BMS APIs and authentication.

uzair114261 opened the pull request:
Added APIs for Bank Management System (BMS), including endpoints for bank creation, listing, and retrieving user account information. Implemented token-based authentication and login API for secure access.

uzair114261 committed:
Feature:Created the API for following:
1. APIs to manage the BMS
2. Token based authentication & API for login.

uzair114261 committed:
Feature:Created the API for following:
1. APIs to delete bank accounts, and update and delete the user account
2. built the routes and views for logout and signup using APIs.

hassan404 commented on accounts/api/serializers.py:
```diff
@@ -0,0 +1,12 @@
+from rest_framework import serializers
```
Dont see django rest framework in requirements.txt

hassan404 commented on banks/api/views.py:
```diff
@@ -0,0 +1,27 @@
+from .pagination import CustomPageNumberPagination
+from rest_framework import generics
+
+class Banks(generics.ListAPIView):
```
```suggestion
class BankListAPIView(generics.ListAPIView):
```

hassan404 commented on banks/api/views.py:
```diff
@@ -0,0 +1,27 @@
+    queryset = Bank.objects.all().order_by('id')
+    serializer_class = BankSerializer
+    pagination_class = CustomPageNumberPagination
+    def list(self, request, *args, **kwargs):
+        items_per_page = request.query_params.get('items_per_page', 10)
+        print(request.user)
+        queryset = self.queryset
+        paginator = self.pagination_class()
+        paginator.page_size = int(items_per_page)
+        page = paginator.paginate_queryset(queryset, request)
+        serializer = BankSerializer(page, many=True)
+        return paginator.get_paginated_response(serializer.data)
```
This is not needed to be overwritten

uzair114261 replied:
Resolved

hassan404 commented on banks/api/urls.py:
```diff
@@ -0,0 +1,7 @@
+from . import  views
+
+urlpatterns = [
+    path('api/list_all_banks', views.Banks.as_view()),
```
```suggestion
    path('api/banks', views.Banks.as_view()),
```

hassan404 commented on banks/api/urls.py:
```diff
@@ -0,0 +1,7 @@
+
+urlpatterns = [
+    path('api/list_all_banks', views.Banks.as_view()),
+    path('api/add_new_bank', views.BankCreateView.as_view())
```
```suggestion
    path('api/banks/create', views.BankCreateView.as_view())
```

hassan404 commented on bank_management_system/urls.py:
```diff
@@ -21,5 +21,9 @@
+    path('api/banks/', include('banks.api.urls')),
+    path('api/auth/', include('auth.api.urls')),
+    path('api/accounts/', include('accounts.api.urls'))
 ]
```
need to have better routes and specify names for each url

uzair114261 replied:
resolved

hassan404 commented on accounts/api/views.py:
```diff
@@ -0,0 +1,94 @@
+
+    def post(self, request):
+        bank_id = request.data.get('bank_id')
+        balance = request.data.get('balance')
+        bank = self.get_bank(bank_id)
+        account = Account.objects.create(
+            bank=bank,
+            user=request.user,
+            balance=balance
+        )
+        data = {
+            "id": account.id,
+            "bank_name": account.bank.bank_name,
+            "branch_name": account.bank.branch_name,
+            "username": account.user.get_full_name(),
+            "balance": account.balance,
+            "bank": account.bank.id,
+            "user": account.user.id
+        }
```
Lets use serializer to handle object creation, listing, update etc

uzair114261 replied:
resolved

hassan404 commented on accounts/api/serializers.py:
```diff
@@ -0,0 +1,12 @@
+from ..models import Account
+
+class AccountSerializer(serializers.ModelSerializer):
+    def get_username(self, obj):
```
Create explicit User class by inheriting from AbstractUser

uzair114261 replied:
resolved

uzair114261 committed:
Feature:Created the API for following:
1. APIs to delete bank accounts, and update and delete the user account
2. built the routes and views for logout and signup using APIs.

hassan404 commented on banks/api/views.py:
```diff
@@ -1,27 +1,38 @@
+
+class BankAPIView(APIView):
+    def post(self, request, *args, **kwargs):
+        serializer = BankSerializer(request.data)
```
need to handle serializer validation and object creation

uzair114261 replied:
resolved

hassan404 commented on banks/api/views.py:
```diff
@@ -1,27 +1,38 @@
 
-class BankCreateView(generics.CreateAPIView):
+# BankList via viewset
+class BankListViewSet(viewsets.ModelViewSet):
```
should be BankViewSet

uzair114261 replied:
resolved

hassan404 commented on banks/api/views.py:
```diff
@@ -1,27 +1,38 @@
-        serializer = BankSerializer(page, many=True)
-        return paginator.get_paginated_response(serializer.data)
 
+class BankActionGenericView(generics.RetrieveAPIView, generics.DestroyAPIView, generics.UpdateAPIView):
```
inherit from composite class

hassan404 commented on accounts/models.py:
```diff
@@ -1,12 +1,14 @@
 
 # Create your models here.
 class Account(models.Model):
     bank = models.ForeignKey(Bank, on_delete=models.CASCADE, related_name='bank')
```
fix related_names everywhere

hassan404 commented on accounts/models.py:
```diff
@@ -1,12 +1,14 @@
     balance = models.PositiveIntegerField()
 
+    class Meta:
+        unique_together = ('user', 'bank')
```
remove this

hassan404 commented on accounts/api/serializers.py:
```diff
@@ -6,7 +6,16 @@ def get_username(self, obj):
         return obj.user.get_full_name()
```
move to User model as property

hassan404 commented on accounts/api/serializers.py:
```diff
@@ -6,7 +6,16 @@ def get_username(self, obj):
         return obj.user.get_full_name()
     bank_name = serializers.ReadOnlyField(source='bank.bank_name', read_only=True)
     branch_name = serializers.ReadOnlyField(source='bank.branch_name', read_only=True)
+    user_id = serializers.ReadOnlyField(source='user.id', read_only=True)
```
make a serializer for User and use here

uzair114261 replied:
resolved

hassan404 commented on accounts/api/serializers.py:
```diff
@@ -6,7 +6,16 @@ def get_username(self, obj):
         return obj.user.get_full_name()
     bank_name = serializers.ReadOnlyField(source='bank.bank_name', read_only=True)
```
use CharField here with read_only True

hassan404 commented on accounts/api/serializers.py:
```diff
@@ -6,7 +6,16 @@ def get_username(self, obj):
+        fields = ['id', 'bank', 'balance', 'bank_name','branch_name','username','user_id']
+
+
+    def update(self, instance, validated_data):
```
this is not needed

hassan404 commented on accounts/api/views.py:
```diff
@@ -1,13 +1,15 @@
-
-class BankAccount(APIView):
+# By APIViews
+class BankAccountAPIView(APIView):
```
AccountAPIView

hassan404 commented on accounts/api/views.py:
```diff
@@ -1,13 +1,15 @@
+class BankAccountAPIView(APIView):
     def get_bank(self, bank_id):
         try:
             return Bank.objects.get(id=bank_id)
```
this is not needed

hassan404 commented on accounts/api/serializers.py:
```diff
@@ -6,7 +6,16 @@ def get_username(self, obj):
         return obj.user.get_full_name()
     bank_name = serializers.ReadOnlyField(source='bank.bank_name', read_only=True)
     branch_name = serializers.ReadOnlyField(source='bank.branch_name', read_only=True)
```
Use Bank Serializer here

uzair114261 replied:
resolved

hassan404 commented on accounts/api/views.py:
```diff
@@ -25,24 +27,21 @@ def get(self, request):
+        account_data = {
+            'bank': bank_id,
+            'user': request.user.id,
+            'balance': balance
```
this should happen through serializer

hassan404 commented on accounts/api/views.py:
```diff
@@ -25,24 +27,21 @@ def get(self, request):
+        else:
+            return Response({
+                "error": serializer.errors
+            },status=status.HTTP_404_NOT_FOUND)
```
400

hassan404 commented on accounts/api/views.py:
```diff
@@ -25,24 +27,21 @@ def get(self, request):
+            },status=status.HTTP_404_NOT_FOUND)
 
     def patch(self, request):
         bank_id = request.data.get('bank_id')
```
pass request.data to serializer and let it handle the validation or input parsing

hassan404 commented on accounts/api/views.py:
```diff
@@ -83,12 +92,38 @@ def delete(self, request):
+class BankAccountGenericView(generics.CreateAPIView, generics.ListAPIView):
+    serializer_class = AccountSerializer
+    def get_queryset(self):
+        return Account.objects.filter(user=self.request.user).select_related('user','bank')
```
lets get user id from kwargs and use that for filtering

hassan404 commented on accounts/api/views.py:
```diff
@@ -83,12 +92,38 @@ def delete(self, request):
+        if serializer.is_valid():
+            serializer.save(user=self.request.user)
+        else:
+            print(serializer.errors)
```
do we need this?

hassan404 commented on accounts/api/views.py:
```diff
@@ -83,12 +92,38 @@ def delete(self, request):
+# By Generic Views
+class AccountActionGenericView(generics.RetrieveAPIView, generics.DestroyAPIView, generics.UpdateAPIView):
     serializer_class = AccountSerializer
     lookup_field = 'bank_id'
```
can just change it to id

uzair114261 committed:
Feature:Created the API with the help of generic classes, ViewSets and APIView.

hassan404 commented on accounts/models.py:
```diff
@@ -2,13 +2,10 @@
 class Account(models.Model):
-    bank = models.ForeignKey(Bank, on_delete=models.CASCADE, related_name='bank')
-    user = models.ForeignKey(User, on_delete=models.CASCADE, related_name='account')
+    bank = models.ForeignKey(Bank, on_delete=models.CASCADE, related_name='bank_account')
+    user = models.ForeignKey(User, on_delete=models.CASCADE, related_name='user_account')
```
since its reverse access of foreign key, the related name should be plural

hassan404 replied:
`accounts` would be a good name for both foreignkeys

uzair114261 replied:
resolved

hassan404 commented on accounts/api/views.py:
```diff
@@ -93,37 +68,30 @@ def delete(self, request):
     serializer_class = AccountSerializer
+    queryset = Account.objects.all()
+
+class ListAccountGenericView(generics.ListAPIView):
```
These are APIViews not "Views"

uzair114261 replied:
resolved

hassan404 commented on accounts/api/views.py:
```diff
@@ -1,38 +1,24 @@
+        print(accounts.query)
+        serializer = AccountDetailSerializer(accounts, many=True)
         return Response({
             "success": True,
```
Why do we need to send "success": True here? Did we implement a case where success can be False?

uzair114261 replied:
resolved

hassan404 commented on accounts/api/serializers.py:
```diff
@@ -1,21 +1,35 @@
-            instance.balance = validated_data['balance']
-            instance.save()
-            return  instance
+        account = Account.objects.create(
```
better to call super().create here and pass validated_data

uzair114261 replied:
resolved

uzair114261 committed:
Fixes: related_names of ForeignKeys are changed and create method of Serializers is updated.
